File: swift/484_FindPermutation.swift
class Solution {
    func findPermutation(_ str: String) -> [Int] {
        var s = str
        var stack: [Int] = [Int](), res: [Int] = [Int]()
        var stringIdx = s.index(s.startIndex, offsetBy: 0)
        var N = s.count
        
        for i in 0...N {
            if i == N || s[stringIdx] == "I" {
                res.append(i+1)
                while !stack.isEmpty { res.append(stack.popLast()!) }
            } else {
                stack.append(i+1)
            }
            if i < N { s.formIndex(after: &stringIdx) }
        }
        return res
    }
}

File: cpp/484_FindPermutation.cpp
class Solution {
public:
    vector<int> findPermutation(string s) {
        vector<int> stack = {}, res = {};
        int N = s.length()+1;
        for (int i = 0; i < N; ++i) {
            if (s[i] == 'D') {
                stack.push_back(i+1);
            } else {
                res.push_back(i+1);
                while (!stack.empty()) {
                    res.push_back(stack.back());
                    stack.pop_back();
                }
            }
        }
        return res;
    }
};
